### DISCUSSION TITLE
Out of range errors for water equation of state #porous_flow #fluid_properties

### DISCUSSION THREAD

**Discussion by rpodgorney** [idaholab/moose#19135] [Q&A Modules: General] [answered]
I have a simulation case where one boundary is practically at the lower pressure and temperature range for the water EOS (1 ATM and 15C) --  the issue--on transient simulation I get out of range errors as the solver iterates.  These stop all meaningful progress of my simulation.  

Has anyone modified the water EOS so that when it goes out of range it DOES NOT return the error but just a warning, and gives the simulation the nearest set of fluid properties?  This requires changing the water EOS to be able to provide fluid properties for conditions that are seemingly below freezing......but this is only a result of the solution process, its not physical in the simulation.

I tried using tabulated fluid properties and manually extended the table to go to low temperatures and pressures--and then changed the range provided in the input file--to no avail......

tagging @andrsd @WilkAndy

**Comment by WilkAndy** [comment]
Also @cpgr will probably be able to say something useful.

I was going to suggest tabulated fluid props.  I wonder why that didn't work.  Did you just not extrapolate enough, or is the physics actually driving you away from the end regions (eg, your simulation keeps trying to reduce temperature or pressure)?   If you use higher pressure or temperature, just as a trial, does MOOSE behave OK?  I wonder whether bounds in the Newton process would help here (usually they just result in crappy convergence)?

**Comment by cpgr** [comment]
Hmmm, they are supposed to throw an exception so that the timestep gets cut, not throw an error and quit.

And strange about tabulated fluid props not working in this case either if you extend the boundaries, as I thought it would just return the end value. 

Is it a `pressure out of range` type error message or something else?

**Comment by rpodgorney** [comment]
Thanks @crgr and @wilkandy  you are right, it does just cut the timestep—which eventually drive the simulation to fail.

It has been a while since I had this issue, I’ll run it again and get a better description, and also test moving the BC away from the EOS min values


--
Robert K. Podgorney, PhD
Chief Computational Scientist
Advanced Scientific Computing Division
Idaho National Laboratory

Office: 208-526-1524<tel:208-526-1524>
email: ***@***.******@***.***>
Statement of Confidentiality
The contents of this e-mail message and any attachments are confidential and are intended solely for addressee. The information may also be legally privileged. This transmission is sent in trust, for the sole purpose of delivery to the intended recipient. If you have received this transmission in error, any use, reproduction or dissemination of this transmission is strictly prohibited. If you are not the intended recipient, please immediately notify the sender by reply e-mail or phone and delete this message and its attachments, if any.


From: Chris Green ***@***.***>
Date: Monday, October 18, 2021 at 3:43 PM
To: idaholab/moose ***@***.***>
Cc: Robert K. Podgorney ***@***.***>, Author ***@***.***>
Subject: [EXTERNAL] Re: [idaholab/moose] Out of range errors for water equation of state #porous_flow #fluid_properties (Discussion #19135)

Hmmm, they are supposed to throw an exception so that the timestep gets cut, not throw an error and quit.

And strange about tabulated fluid props not working in this case either if you extend the boundaries, as I thought it would just return the end value.

Is it a pressure out of range type error message or something else?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fidaholab%2Fmoose%2Fdiscussions%2F19135%23discussioncomment-1498795&data=04%7C01%7Crobert.podgorney%40inl.gov%7C66af2fdadfc4452328d708d9928056a1%7C4cf464b7869a42368da2a98566485554%7C0%7C0%7C637701902178744576%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=XUhtrIUrR%2FxYO%2B%2BdRBF3a4PLD2og5JTlFjz6oULd7EQ%3D&reserved=0>, or unsubscribe<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FACWGDT34JCP4HBDJS7TE45LUHSIIPANCNFSM5GHCJRTA&data=04%7C01%7Crobert.podgorney%40inl.gov%7C66af2fdadfc4452328d708d9928056a1%7C4cf464b7869a42368da2a98566485554%7C0%7C0%7C637701902178754535%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=NCnm4iHZUmNJJI%2F8Sd7Ou5NnerE5LdBr7vmLJLie3eY%3D&reserved=0>.
Triage notifications on the go with GitHub Mobile for iOS<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fapps.apple.com%2Fapp%2Fapple-store%2Fid1477376905%3Fct%3Dnotification-email%26mt%3D8%26pt%3D524675&data=04%7C01%7Crobert.podgorney%40inl.gov%7C66af2fdadfc4452328d708d9928056a1%7C4cf464b7869a42368da2a98566485554%7C0%7C0%7C637701902178754535%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=pQzJcNy1NrpKVX04YGpHJRdfCbv3Lhaa%2FeJJNJhnGPo%3D&reserved=0> or Android<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fplay.google.com%2Fstore%2Fapps%2Fdetails%3Fid%3Dcom.github.android%26referrer%3Dutm_campaign%253Dnotification-email%2526utm_medium%253Demail%2526utm_source%253Dgithub&data=04%7C01%7Crobert.podgorney%40inl.gov%7C66af2fdadfc4452328d708d9928056a1%7C4cf464b7869a42368da2a98566485554%7C0%7C0%7C637701902178754535%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=o6PfU8EveYLmqtdcaA%2FKYwzQWwej%2FnuXpTRyZYUM7x4%3D&reserved=0>.

**Comment by andrsd** [comment]
IMO if you are outside of the range of validity you have 2 options: (1) extend the range of validity of the fluid property class, (2) change your model so you are within the range of validity.

Regarding NaNs: I would highly recommend that `Water97FluidProperties` class gains the functionality of `NaNInterface`. Look at [this PR](https://github.com/idaholab/moose/pull/19183/files) we did today what I mean by that. It is a trivial change that gives users option to threat NaNs as errors, warnings or ignore them.

If anybody decides to implement it, feel free to request a review from me. Obviously, other classes in the FP module would benefit from this as well, not only `Water97FluidProperties` :-)

**Reply by cpgr** [reply]
Thanks for introducing `NaNInterface`. I'll get around to including it in the classes when I get a chance

**Reply by WilkAndy** [reply]
> Thanks for introducing `NaNInterface`. I'll get around to including it in the classes when I get a chance

Cam we use this within PorousFlow, @cprg?  Off the top of my head, i can't think of anywhere especially appropriate.  Eg, the relperm curves have bounds, but i think it's an error if somehow the saturation goes outside those.

**Reply by cpgr** [reply]
Not sure. I think we return the bounds of ref perm if saturation goes outside range, rather than throw an exception.

**Reply by andrsd** [reply]
Just for the record `NaNInterface` can be used with any object, not just `FluidProperties` - I realized :-) Maybe, it should go into framework, really...

**Comment by rpodgorney** [comment, marked as answer]
Thanks everyone---I explored this in detail yesterday.  The simplest answer was that the problems I was have were due to user error.  I am not able to extend the tabulated range and run my problems just fine.
